src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs

```diff
@@ -58,12 +66,14 @@ public async Task<MsalTokenResponse> GetTokenAsync(Uri endPoint, RequestContext
             return await ExecuteRequestAsync<MsalTokenResponse>(endPoint, HttpMethod.Post, requestContext).ConfigureAwait(false);
         }
 
-
-
-        internal async Task<T> ExecuteRequestAsync<T>(Uri endPoint, HttpMethod method, RequestContext requestContext, bool expectErrorsOn200OK = false)
+        internal async Task<T> ExecuteRequestAsync<T>(Uri endPoint, HttpMethod method, RequestContext requestContext, bool expectErrorsOn200OK = false, bool respondToDeviceAuthChallenge = true)
         {
             bool addCorrelationId = requestContext != null && !string.IsNullOrEmpty(requestContext.Logger.CorrelationId.ToString());
-            AddCommonHeaders(requestContext, addCorrelationId);
+            //No need to re-add headers for the replayed PKeyAuth request
+            if (respondToDeviceAuthChallenge)
+            {
+                AddCommonHeaders(requestContext, addCorrelationId);
```

Comment:
@henrik-me @jmprieur  - please clarify if we want to do PKAuth every time? This results in a pretty big pref impact.

This seems to effectively add an extra HTTP call to the `/token` endpoint. Not to mention all the crypto operations which also take time. I don't know if @trwalke measured the perf impact of PKAuth - perhaps he can add some numbers.

Reply:
Wanted to clarify, this line is adding a predefined pkey auth header to the request that lets the STS know that the client is capable of handling PKeyAuth. The only time MSAL will actually perform the challenge is when authentication fails with a very specific error that contains the PKeyAuth challenge data. This will be the only time MSAL will sign the challenge and then replay the request. ADAL is using the same logic for this here, however, it is still up for debate if we want to go in a different direction.

Reply:
Let's add some headers, they are only used for telemetry purposes, so it's better to have them.

Reply:
Still active comment - please leave those headers there.